dashboard/cybershake_progress.py

```diff
@@ -0,0 +1,373 @@
+
+import qcore.simulation_structure as sim_struct
+import qcore.constants as const
+from estimation.estimate_cybershake import main as est_cybershake
+
+# The process types that are used for progress tracking
+PROCESS_TYPES = [
+    const.ProcessType.EMOD3D.str_value,
+    const.ProcessType.HF.str_value,
+    const.ProcessType.BB.str_value,
+]
+
+EST_CORE_HOURS_COL = "est_core_hours"
+ACT_CORE_HOURS_COL = "act_core_hours"
+
+COMPLETED_R_COUNT_COL = "completed_r_count"
+R_COUNT_COL = "r_count"
+MISSING_DATA_FLAG_COL = "missing_data"
+
+
+def get_chours(json_file, proc_types: List[str]):
+    """Gets the core hours for the specified process types (i.e. EMOD3D, HF, BB, ...)
+    for the specified simulation metadata json log file.
+    """
+    with open(json_file, "r") as f:
+        data = json.load(f)
+
+    sim_name = data.get(const.MetadataField.sim_name.value)
+    core_hours = []
+    for proc_type in proc_types:
+        if proc_type in data.keys():
+            core_hours.append(
+                data[proc_type].get(const.MetadataField.run_time.value, np.nan)
+                * data[proc_type].get(const.MetadataField.n_cores.value, np.nan)
+            )
+        else:
+            core_hours.append(np.nan)
+
+    return sim_name, core_hours
+
+
+def get_chours_used(fault_dirs: Iterable[str]):
+    """Returns a dataframe containing the core hours used for each of the
+    specified faults"""
+    faults, core_hours, completed_r_counts, missing_data = [], [], [], []
+    for fault in fault_dirs:
+        fault_name = os.path.basename(fault)
+
+        cur_fault_chours = np.asarray(
+            [
+                get_chours(json_file, PROCESS_TYPES)[1]
+                for json_file in glob.glob(
+                    os.path.join(fault, "**", const.METADATA_LOG_FILENAME),
+                    recursive=True,
+                )
+            ]
+        ).reshape(-1, len(PROCESS_TYPES))
+
+        missing_data.append(np.any(np.isnan(cur_fault_chours)))
+
+        completed_r_counts.append(
+            np.count_nonzero(~np.any(np.isnan(cur_fault_chours), axis=1))
+        )
+
+        faults.append(fault_name)
+        if cur_fault_chours.shape[0] > 0:
+            core_hours.append(np.nansum(cur_fault_chours, axis=0))
+        else:
+            core_hours.append([np.nan, np.nan, np.nan])
+
+    df = pd.DataFrame(index=faults, data=core_hours, columns=PROCESS_TYPES)
+    df[COMPLETED_R_COUNT_COL] = completed_r_counts
+    df[MISSING_DATA_FLAG_COL] = missing_data
+
+    return df
+
+
+def get_faults_and_r_count(cybershake_list: str):
+    """Gets the fault names and number of realisations from a cybershake fault list."""
+    with open(cybershake_list, "r") as f:
+        lines = f.readlines()
+
+    fault_names, r_counts = [], []
+    for ix, line in enumerate(lines):
+        line_l = line.split(" ")
+        fault_names.append(line_l[0].strip())
+
+        try:
+            r_counts.append(int(line_l[1].strip().rstrip("r")))
```

Comment:
We need to standardise the 'r'

Reply:
https://github.com/ucgmsim/slurm_gm_workflow/blob/master/scripts/cybershake/install_cybershake.sh#L28

NKR are all currently valid realisation specifications. A conversation about wether we are deprecating the old ones is a worthwhile.

Reply:
Have added "n" and "k"